### ISSUE TITLE
Tutorial 5: Memory and POMDPs with recurrent PPO

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#311]
Child of #303 (tutorial series). Add `docs/tutorials/05-memory-pomdps.md`: recurrent (LSTM) PPO on FlickeringCartPole — when an LSTM earns its cost, recurrent rollouts, hidden-state handling, and the memoryless-MLP contrast.

**Requirements (match tutorials 1–2):**
- Single Markdown file with complete copy-paste code.
- Code CI-enforced as a doc-test via `src/tutorials.rs` (`tutorial_05_memory` submodule).
- Builds on tutorial 2's PPO loop; needs on-policy but not the off-policy tutorials.
- Voice matches GETTING_STARTED.md; explain *why*.
- Update `docs/tutorials/README.md` status row.
- Full test suite + clippy -D warnings + fmt + doc build clean.

Reference the packaged `recurrent_ppo_flickering_cartpole` example and docs/RECURRENT_POLICY_DESIGN.md.

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating this issue per user request. Adding technical context, exact API surface, doc-test wrapper guidance, CI-safe step budget, concrete acceptance criteria, implementation guidance, and a scope guard.

---

### Context: What Exists on main Today

Tutorials 1–3 have landed. The tutorial infrastructure is:

- Each tutorial is a single Markdown file under `docs/tutorials/`.
- Every `rust` code block in a landed tutorial is compiled and run as a doc-test by `cargo test --features training` via `#[doc = include_str!(...)]` in `src/tutorials.rs`.
- No `no_run` or `ignore` directive is used; every code block executes.
- `type Backend = Autodiff<NdArray<f32>>;` is used across all three landed tutorials.

The full recurrent stack (LSTM policy, recurrent rollout buffer, recurrent PPO trainer, FlickeringCartPole environment) is all on `main`. Tutorial 5 is a documentation issue — the implementation is complete and is the source of truth.

Reference example: `examples/games/cartpole/recurrent_ppo_flickering_cartpole.rs`. The tutorial's code block is a stripped-down, pedagogical version of this example.

---

### Technical Context: Exact API Surface

#### FlickeringCartPole (`thrust_rl::env::FlickeringCartPole`)

Not in the prelude; requires an explicit import from `thrust_rl::env::flickering_cartpole::FlickeringCartPole` or `thrust_rl::env::FlickeringCartPole`.

```rust
FlickeringCartPole::new()                                        // entropy-seeded, p=0.5
FlickeringCartPole::with_seed_and_probability(seed: u64, p: f64) // reproducible; panics if p ∉ [0,1]
```

**Burst-mode constructor** (PR #318, merged today — stricter assert):
```rust
FlickeringCartPole::with_seed_probability_and_burst(seed, flicker_prob, burst_len)
// Panics if:
//   flicker_prob ∉ (0, 1)   (open interval — both states must be reachable)
//   burst_len < 1.0
//   flicker_prob > burst_len / (burst_len + 1)
// Example valid: p=0.5, burst_len=4 → max_p = 4/5 = 0.80 → 0.5 ≤ 0.80 ✓
// Example invalid: p=0.9, burst_len=4 → max_p = 0.80 → 0.9 > 0.80 ✗ (panic)
```

Tutorial 5 uses **only** the non-burst constructors. The burst example is out of scope (reference in "Going further").

Observation space: 4-D `[x, x_dot, theta, theta_dot]`. A blanked frame is `[0, 0, 0, 0]` — same shape, zeroed contents.

#### LSTM Policy (`thrust_rl::policy::lstm`)

```rust
use thrust_rl::policy::lstm::{LstmBurnConfig, LstmBurnPolicy};

let config = LstmBurnConfig { hidden_dim: 64, ..Default::default() }.with_seed(SEED);
let policy = LstmBurnPolicy::<Backend>::with_config(obs_dim, action_dim, config, &device);

// Rollout (single-step):
let (logits, values, new_state) = policy.forward_step(
    obs_t: Tensor<B, 2>,           // [N_env, obs_dim]
    state:  Option<LstmState<B, 2>> // None → zeros; Some(s) → warm-start
);
// → logits: [N_env, action_dim],  values: [N_env],  new_state: LstmState<B, 2>

// Training (rank-3 forward):
let (log_probs, entropy, values) = policy.evaluate_sequences(
    obs_seq:        Tensor<B, 3>,      // [N_env, T, obs_dim]
    actions:        Tensor<B, 2, Int>, // [N_env, T]
    episode_starts: Tensor<B, 2>,      // [N_env, T], 1.0 where obs[t] begins a new episode
);
// → each output: [N_env, T]
```

`LstmBurnPolicy` is NOT in the prelude; needs explicit import.

#### RecurrentRolloutBuffer (`thrust_rl::buffer::rollout::RecurrentRolloutBuffer`)

```rust
// Construction
let mut buffer = RecurrentRolloutBuffer::new(num_steps, num_envs, obs_dim, hidden_dim);

// Per-step, per-env additions (inside rollout loop):
buffer.add(step, env, obs: &[f32], action: i64, reward: f32,
           value: f32, log_prob: f32, terminated: bool, truncated: bool);
buffer.add_recurrent_state(step, env, h: &[f32], c: &[f32]); // NEW vs feedforward

// Between rollout iterations:
buffer.seed_warm_start(last_h: &[f32], last_c: &[f32], last_done: &[bool]);
buffer.reset();            // clears transition data; preserves (h,c) warm-start carry
buffer.compute_advantages(last_values: &[f32], gamma: f32, gae_lambda: f32);
```

NOT in the prelude; import as `thrust_rl::buffer::rollout::RecurrentRolloutBuffer`.

GAE / state-reset asymmetry (do not homogenize): GAE resets on `terminated` only; `episode_starts` resets on `terminated || truncated`. The buffer handles this correctly internally — the tutorial should explain the conceptual distinction without re-implementing it.

#### RecurrentPPOTrainer (`thrust_rl::train::ppo::RecurrentPPOTrainer`)

```rust
use thrust_rl::train::ppo::RecurrentPPOTrainer;

let mut trainer = RecurrentPPOTrainer::new(
    config:    PPOConfig,
    policy:    LstmBurnPolicy<B>,
    optimizer: BurnOptimizer<B, LstmBurnPolicy<B>, O>,
    device:    B::Device,   // ← NEW argument vs PPOTrainerBurn (required for rank-3 batch materialization)
)?;

trainer.policy() -> &P      // borrow policy during rollout
trainer.set_learning_rate(lr: f64)           // LR annealing support
trainer.increment_episodes(n: usize)         // caller tracks episode count

trainer.train_step(
    &buffer,
    envs_per_minibatch: usize,               // whole trajectories per minibatch (not loose timesteps)
    |policy, obs_seq: Tensor<B,3>, actions: Tensor<B,2,Int>, episode_starts: Tensor<B,2>|
        -> (log_probs: Tensor<B,2>, entropy: Tensor<B,2>, values: Tensor<B,2>)
) -> Result<TrainingStats>
```

`RecurrentPPOTrainer` is NOT in the prelude; needs explicit import.

#### Full import block (matches the packaged example)

```rust
use burn::{
    backend::Autodiff,
    nn::LstmState,
    optim::AdamConfig,
    tensor::{Int, Tensor, TensorData, activation},
};
use rand::{Rng, SeedableRng, rngs::StdRng};
use thrust_rl::{
    buffer::rollout::RecurrentRolloutBuffer,
    env::{Environment, SpaceType, flickering_cartpole::FlickeringCartPole, pool::EnvPool},
    policy::{
        lstm::{LstmBurnConfig, LstmBurnPolicy},
        mlp::{BurnActivation, MlpBurnConfig, MlpBurnPolicy},
    },
    train::{
        optimizer::BurnOptimizer,
        ppo::{PPOConfig, PPOTrainerBurn, RecurrentPPOTrainer},
    },
};
type InnerBackend = burn::backend::NdArray<f32>;
type Backend = Autodiff<InnerBackend>;
```

#### Observation normalization (load-bearing)

The tutorial code block must inline a private `ObsNormalizer` helper (defined as a `struct` at the bottom of the code block, exactly like the example). The library's `RunningMeanStd` in `thrust_rl::utils::normalize` does NOT have the "skip blanked frames" special case.

Why this is load-bearing: CartPole's raw observations are tiny (initial perturbations ~0.05). The LSTM's tanh gates squash such small inputs to near-zero features, starving both heads of gradient signal. A blanked frame is exactly all-zeros — a state that CartPole physics never produces — and must be passed through unchanged so the "no observation" signal stays a clean zero for both policies. Folding zeros into the running mean would pollute the normalizer statistics.

---

### CI Step Budget

| Setting | CI doc-test | Real run |
|---|---|---|
| `NUM_ENVS` | 4 | 16 |
| `NUM_STEPS` | 32 | 128 |
| `TOTAL_TIMESTEPS` | 4_096 | 500_000 |
| `HIDDEN_DIM` | 64 | 64 |
| `LSTM_LR` | 1.5e-3 | 1.5e-3 |
| `MLP_LR` | 3e-4 | 3e-4 |
| `GAMMA` | 0.99 | 0.99 |
| `GAE_LAMBDA` | 0.95 | 0.95 |
| `N_EPOCHS` | 4 | 4 |
| `ENVS_PER_MINIBATCH` | 2 | 4 |
| `REWARD_SCALE` | 0.02 | 0.02 |

`TOTAL_TIMESTEPS = 4_096` → `4096 / (4 * 32) = 32 PPO updates`. Inline comment: `// tiny for CI; use ~500_000 for the real run` (matches the prose instruction to run the packaged example for the full run).

PPO config flags that are load-bearing for the recurrent trunk (must be present in the tutorial code, not just mentioned in prose):
- `clip_range_vf(0.2)` — load-bearing with reward scaling; without it, the value clip freezes the critic
- `max_grad_norm(0.5)` — load-bearing; value-loss spikes otherwise wipe out policy features in the LSTM trunk
- `reward_scale` applied on `buffer.add(…, reward * REWARD_SCALE, …)` — not a PPOConfig field; inline scaling

---

### Tutorial Structure Outline

Follow the tutorials 2–3 voice: each section explains the *why* before the code.

1. **Header block** — "You will: / Prerequisites: Tutorial 2 only / Time: ~20 minutes"
2. **The problem** — What a POMDP is; why CartPole-v1 (which has full state) is NOT a POMDP; what flickering does
3. **Why velocity-masking didn't work** — the documented negative result (`MaskedCartPole`); why flickering closes the loophole; citation of Hausknecht & Stone (2015)
4. **When an LSTM earns its cost** — the cost is extra complexity and the warm-start bookkeeping; when does memory pay off vs. a feedforward baseline?
5. **How the LSTM handles a rollout** — `forward_step` threads `(h, c)` across steps; on episode end, zero the hidden state; no warm-start across rollout windows in the simple version (BPTT truncation)
6. **Recurrent training vs. feedforward training** — the key differences: `RecurrentRolloutBuffer` + `add_recurrent_state`, `RecurrentPPOTrainer` takes `device`, `evaluate_sequences` not `evaluate`, `envs_per_minibatch` not `batch_size`
7. **The code block** — complete copy-paste code with both LSTM and MLP arms
8. **Reading the output** — what to look for; `best_mean` (peak of moving average) vs. `final_mean` (end-of-run, potentially noisier)
9. **Scope guard / going further** — i.i.d. flickering only in this tutorial; burst mode + wider ablation in `recurrent_ppo_burst_flickering_cartpole`; the full design rationale in `docs/RECURRENT_POLICY_DESIGN.md`

---

### Implementation Guidance

1. **Start from the packaged example** (`examples/games/cartpole/recurrent_ppo_flickering_cartpole.rs`). The tutorial code block is a pedagogical simplification of `run_lstm()` + `run_feedforward()`, with CI-sized constants substituted in.

2. **Inline `ObsNormalizer`** near the end of the code block, before the first call that uses it. The struct is private to the code block; no import needed.

3. **Wire `src/tutorials.rs`** by adding after the tutorial 3 entry:
   ```rust
   /// Tutorial 5 — Memory and POMDPs with recurrent PPO.
   ///
   /// See the rendered Markdown at
   /// [`docs/tutorials/05-memory-pomdps.md`](https://github.com/rjwalters/thrust/blob/main/docs/tutorials/05-memory-pomdps.md).
   #[doc = include_str!("../docs/tutorials/05-memory-pomdps.md")]
   pub mod tutorial_05_memory {}
   ```
   Also add a bullet in the module-level doc comment listing the new tutorial.

4. **Update `docs/tutorials/README.md`** — change the tutorial 5 status row from "🚧 Planned ([#311]...)" to "✅ Landed (PR #NNN)". This is a one-row edit; include the actual PR number.

5. **`evaluate_sequences` in the training closure** — the `evaluate_fn` passed to `trainer.train_step` should call `policy.evaluate_sequences(obs_seq, actions, episode_starts)` directly, since `LstmBurnPolicy` implements this as its own method (unlike the feedforward trainer, which takes a generic closure typed to rank-2 tensors).

---

### Scope Guard

This tutorial covers **i.i.d. flickering** (the Hausknecht & Stone 2015 protocol) only.

Out of scope for tutorial 5:
- Burst-structured (correlated) dropout — demonstrated in `recurrent_ppo_burst_flickering_cartpole`; the burst constructor has a stricter assert (`flicker_prob <= burst_len / (burst_len + 1)`, from PR #318) and adds complexity that belongs in the example, not the introductory tutorial
- `MaskedCartPole` — covered as a documented negative result in prose, not in code
- `recurrent_ppo_masked_cartpole` example
- T-Maze
- Strategy B (fixed-length subsequences) — deferred per the design note
- Multi-env rollout warm-starting beyond what the code block already demonstrates

---

## Acceptance Criteria

- [ ] `docs/tutorials/05-memory-pomdps.md` created with:
  - [ ] Header block naming Tutorial 2 as the sole prerequisite
  - [ ] Explanation of POMDPs, why flickering makes memory load-bearing by construction
  - [ ] Negative-result section: why velocity-masking (MaskedCartPole) did NOT work and what that teaches
  - [ ] Explanation of what the LSTM costs vs. a feedforward policy and when it pays off
  - [ ] Complete single `rust` code block (no `no_run`) that compiles and runs as a doc-test
  - [ ] Code block uses `TOTAL_TIMESTEPS = 4_096` (CI budget) with comment to increase for a real run
  - [ ] Code block runs both LSTM and MLP arms and prints the contrast
  - [ ] Code block inlines `ObsNormalizer` with the "skip blanked frames" logic
  - [ ] "Going further" section pointing to `recurrent_ppo_flickering_cartpole` example and `docs/RECURRENT_POLICY_DESIGN.md`
  - [ ] Voice matches tutorials 1–3: explains *why* before showing *how*

- [ ] `src/tutorials.rs` updated:
  - [ ] `tutorial_05_memory` submodule added with `#[doc = include_str!("../docs/tutorials/05-memory-pomdps.md")]`
  - [ ] Module-level doc comment updated with bullet for tutorial 5

- [ ] `docs/tutorials/README.md` updated:
  - [ ] Tutorial 5 row status updated to "✅ Landed" with PR link

- [ ] CI passes:
  - [ ] `cargo test --features training` (doc-test runs and does not panic)
  - [ ] `cargo clippy --features training -- -D warnings` (no warnings)
  - [ ] `cargo fmt --check` (formatted)
  - [ ] `cargo doc --features training` (no rustdoc warnings)

---

## Test Plan

- [ ] Manual verification: `cargo test --features training --doc tutorials::tutorial_05_memory` completes without error in < 60 seconds on a laptop CPU
- [ ] Automated tests: the doc-test in `docs/tutorials/05-memory-pomdps.md` IS the automated test; no additional unit test file needed
- [ ] Edge case: verify `cargo test --features training` still passes for tutorials 1–3 (regression guard on `src/tutorials.rs` wiring)
- [ ] Style check: read the rendered Markdown on GitHub and verify the code block syntax-highlights correctly and all prose links resolve

---

## Affected Files

- `docs/tutorials/05-memory-pomdps.md` — **new file, to be created by builder** (does not exist on `origin/main`)
- `src/tutorials.rs` — add `tutorial_05_memory` submodule and update module-level doc comment (exists on `origin/main`)
- `docs/tutorials/README.md` — update tutorial 5 status row (exists on `origin/main`)

Pre-flight verification: `src/tutorials.rs` and `docs/tutorials/README.md` confirmed present on `origin/main`. `docs/tutorials/05-memory-pomdps.md` is a net-new file; builder creates it.
